### ISSUE TITLE
Support some sort of cross-browser GC trigger

### ISSUE THREAD

**Issue by jdm** [web-platform-tests/wpt#7899]
#7887 adds a mechanism that is Blink-only for some tests, and https://github.com/w3c/web-platform-tests/blame/master/streams/resources/test-utils.js contains one for Blink and Webkit. It's clear there is a desire to write tests that rely on GC behaviour; it would be useful to find ways to support that.

**Comment by mkruisselbrink** [comment]
Related to this would of course be to actually run the browser such that triggering GC works.

**Comment by foolip** [comment]
Strawperson: move what's in test-utils.js into /common/gc.js and see what happens.

@jgraham, are there any bad things you'd expect to happen?

**Comment by JKereliuk** [comment]
Are there any current manual tests that need this?

**Comment by foolip** [comment]
The existing tests won't be manual test for the most part, they're just tests that define a `gc()` method and try different kinds of things to trigger a GC.

**Comment by jgraham** [comment]
I'm pretty worried about exposing gc behaviour in particular. We have successfully made the precise gc semantics an implementation detail, and I don't want tests to be the blocker to people improving their js engines in the future. In particular I don't know that all browsers are currently able to force a complete synchronous gc at a specific time, so I imagine it wouldn't be possible to implement this `gc()` function cross-browser.

**Comment by foolip** [comment]
Right, in Blink it's not possible to force a precise GC from scripts, that has to happen at the top of the event loop. So such an API would have to be async, I suppose, or not exist at all.

**Comment by foolip** [comment]
As an input for prioritization, GC came up in both replies to my [Blockers to using web-platform-tests?](https://groups.google.com/a/chromium.org/d/msg/blink-dev/mVxq1zLYvYg/qcGjULW4CQAJ) thread. I didn't look at the WebRTC tests, but @atotic confirms that the ResizeObserver case could be serves by a GC helper, sync or async.

**Comment by jgraham** [comment]
@bzbarsky You seem like a good person to comment from a gecko point of view :)

**Comment by bzbarsky** [comment]
In gecko, there is no deterministic way to force "all of GC" to run.  In particular, the JS GC and cycle collector can both free up stuff that then causes the other to find more garbage.  In theory, this can go arbitrarily deep, depending on the cycle collection implementations.

In practice, Gecko tests that try to test GC stuff make certain assumptions about the structure of the graph and either do a GC+CC once or three times or something.

There are further complications if there are multiple toplevel browsing contexts involved, and even further ones if multiple units of related browsing contexts involved, and some of the things around that are in flux.  If we need anything resembling reliable GC across all renderer processes (e.g. to test some sort of MessageChannel thing), I'm 99% sure Gecko has nothing like that right now.

Anyway, if we can figure out some sort of approximate semantics for the GC trigger, and don't rely on GC being too precise, this is probably viable.

**Comment by atotic** [comment]
There are 2 assertions tests might be interested in after the gc() call:
1) assert that the object is still alive
2) that the object has been garbage collected

ResizeObserver asserts (1), object is still alive. It asserts that ResizeObserver continues broadcasting after JS reference to the observer has been dropped.

According to Boris, gc() with full collection guarantees cannot be implemented in Firefox. They have something that's more like a "best effort" gc. What this means is:
1) tests with type 1 asserts will always pass. They might pass when they should have failed.
2) tests with type 2 asserts will be flaky.

If I was w-p-t, I'd implement "best effort" gc extension, document that type 2) asserts are not allowed, and reject any tests that try making such assertions. This would allow gc enabled tests to land, and get widest possible coverage.

**Comment by bzbarsky** [comment]
> According to Boris, gc() with full collection guarantees cannot be implemented in Firefox.

It's not just Firefox, fwiw.  WebKit's GC, according to the recent post at <https://webkit.org/blog/7122/introducing-riptide-webkits-retreating-wavefront-concurrent-garbage-collector/>, includes a conservative stack-and-register scanner.  That means that it can never _guarantee_ that something gets collected; it's possible that some random integer on the stack looks like a pointer to your object and keep it alive.

I'm a little curious how you write a "type 1" assertion, given that collection of objects is not supposed to have observable side-effects.  That is, you can test for things like whether the ResizeObserver keeps broadcasting, but if it _doesn't_ that tells you nothing about whether it's still _alive_.  On the other hand, what you care about is whether it's broadcasting, not whether it's alive, right?

**Comment by atotic** [comment]
> I'm a little curious how you write a "type 1" assertion, given that collection of objects is not supposed to have observable side-effects. That is, you can test for things like whether the ResizeObserver keeps broadcasting, but if it doesn't that tells you nothing about whether it's still alive. On the other hand, what you care about is whether it's broadcasting, not whether it's alive, right?

Correct. The test is whether ResizeObserver keeps broadcasting. The reason for this test is that some Chrome GC changes caused RO to be collected and stop broadcasting when JS reference count went to 0. I created a test to make sure it does not happen again.

**Comment by alijuma** [comment]
This has priority:roadmap but hasn't been updated in a while. Given the comments so far, does it still make sense to try to have a cross-browser GC trigger of some sort? And if so, who's a good owner for this?

**Comment by foolip** [comment]
@jgraham, ping on the notion of having testdriver.js APIs not backed by WebDriver. This idea has also come up in the discussion around https://github.com/web-platform-tests/wpt/pull/10746, where we could have a `test_driver.get_noise_stream` with a default implementation like this:
https://github.com/web-platform-tests/wpt/blob/36c9fdd4ef36c289f416a16707074157d818ea7c/webrtc/RTCPeerConnection-helper.js#L412

But where implementers can overrride it. WDYT?

cc @jugglinmike @youennf

**Comment by jgraham** [comment]
I think if we start to do this, particuarly in the gc case where we can't offer a meaningful default implementation, we actually need to write a spec on what the API guarantees. For testdriver-backed-by-WebDriver that spec is implicitly the WebDriver spec so a custom implementation can figure out if it's compatible by looking at that spec and ensuring it has the same behaviour. For other things I propose we require a similar level of detail.

This dovetails a little about thoughts I have for changing API surface in general (across the project), which are a little out of scope for this bug, but in essence now that there are more people contributing to the project I thin kit would make sense to adopt something more like an RFC system so that people are able to keep up with changes and ensure that decisions are being made on the basis of public discussion.

**Comment by foolip** [comment]
This sounds like a similar bar as what we've asked for with test-only APIs, and I agree that we need clarity on what these APIs do. But, I think that documenting it as part of the wpt project should suffice for some cases. As an example, if we have a `get_media_stream` which only guarantees that there's *some* non-static video and non-silent audio tracks, then just put that in the `testdriver.js` API documentation.

In the case of GC, there strictly speaking can't be any guarantees at all, because GC shouldn't be observable. And yet we have plenty of tests that make some effort to trigger GC and are more likely to reveal bugs in GC is triggered. So in this case, I would suggest a typical default implementation of creating a lot of objects + setTimeout, and documenting it as a best effort. Even https://tc39.github.io/ecma262/ doesn't say anything normative about garbage collection, so I don't know what spec-like language we could use here.

On the RFC thing, I'd be up for something like that. Do you want to file a `policy` issue so we can discuss together with other governance stuff, maybe at TPAC?

**Comment by jgraham** [comment]
> As an example, if we have a get_media_stream which only guarantees that there's some non-static video and non-silent audio tracks, then just put that in the testdriver.js API documentation.

In that case the documentation becomes a de-facto spec. I think there's a real concern here that if you make the definitions too loose (e.g. just guarantee that there's *something* in the audio track) a test written against one implementation may  not do anything meangful in a different implemenation. If as a result the test happens to FAIL then someone will likely notice and do something about it. If it happens to PASS we will most likely miss the interop issue.

On the specifc topic of gc, I agree that there's nothing we can do to guarantee specific results. But even a best-effort function has to have certain agreed properties (e.g. is it synchronous or async? Do we guarantee that if a gc will happen ir will have happened when the function returns or the promise resolves, or may there be more activity in the background). Knowing what these properties should be depends on getting feedback from all relevant parties about what would work in their implementations, and so it makes sense to document what the conclusions are.

I'll make an issue for the RFC process. I was indeed waiting to talk about it and other govenance topics at TPAC because that seems like the place where we will have most of the relevant parties present.

**Comment by foolip** [comment]
Yes, certainly we need to document the timing of a gc method. I'd suggest that it returns a promise, which is resolved after an effort has been made to trigger GC. We'd have to talk JS engine people about what a good default implementation would be, but maybe something like setTimeout+generate garbage+setTimeout+resolve promise.

**Comment by syg** [comment]
Resurrecting discussion here on @domenic's suggestion. JS WeakRefs are close to shipping and there are some HTML integration tests that would be good to add. It is impossible to write `WeakRef` or `FinalizationRegistry` tests without ability to trigger gc. Currently there are some [tests in flight](https://chromium-review.googlesource.com/c/chromium/src/+/2145160) in Chrome that are Chrome-only.

Also cc @rwaldron for context on what test262 did here.

**Comment by stephenmcgruer** [comment]
@syg can you summarize what guarantees or functionality JS WeakRefs provide, with regards to the HTML integration?

From my side, summarizing what I take away from reading this thread:

GC semantics are in general an implementation detail. This includes:
* What objects get collected during any given garbage collection 'run'.
* Whether the collection can be synchronous or not.

Things atotic noted as possibly desirable to test:
i. assert an object is still alive (has live references)
ii. assert an object has been garbage collected

(i) was noted as more feasible than (ii).

Seems we have at least Blink and Gecko interested in doing something 'not too precise'. Likely an asynchronous call, that returns a promise that resolves when an effort has been made to do garbage collection. Which is also clearly documented to guarantee nothing.

We now have an [RFC process](https://github.com/web-platform-tests/rfcs/blob/master/README.md) where something like this could be hammered out; historically the best way to do that is to post an RFC pull request that has some 'reasonable' (in your opinion) idea and let folks argue it out, changing the PR as necessary. @syg would you be interested in drafting an RFC for this API?

**Comment by syg** [comment]
@stephenmcgruer Sure, I can provide background on the HTML integration parts:

### `WeakRef` 

- When a `weakRef.deref()` observes a non-`null` object (that is, a live referent), that referent is guaranteed to stay live until the end of the event loop. For HTML, that means until just after the current task's microtask checkpoint.

### `FinalizationRegistry`

`FinalizationRegistry` is a way to register objects to observe their post-mortem finalizations. In the following example, `myCallback` would be called with `"holdings 1"` at some point after `someObject1` is GCed, and with `"holdings 2"` at some point after `someObject2` is GCed.

```javascript
let fr = new FinalizationRegistry(function myCallback(holdings) { /* do something */ });
fr.register(someObject1, "holdings 1");
fr.register(someObject2, "holdings 2");
```

There are no guarantees on when and if these callbacks would occur, but if they _do_ occur, there are the following guarantees in HTML:

- The finalization callbacks are run in a task.
- There is a microtask checkpoint after each batch of finalization callbacks in a task -- it's up to the implementation to decide how to batch.
- Errors thrown from within the finalization callback trigger `window.onerror`.

### Testing

Tests of the shape "if the finalizer ran, then the following properties must hold" can be written for the above.  We can't guarantee that the finalizers run, but we can significantly increase its chances with a GC hook. 

Thanks for the pointer to the RFC process. I won't have cycles to do an RFC but perhaps @rwaldron might be able to. Bocoup is helping improve the WeakRef test coverage for the HTML integration part.

**Comment by stephenmcgruer** [comment]
Thanks for the great overview! 

> Thanks for the pointer to the RFC process. I won't have cycles to do an RFC but perhaps @rwaldron might be able to. Bocoup is helping improve the WeakRef test coverage for the HTML integration part.

Ok; I'm happy to help anyone who is working on making an RFC (by looking at drafts, suggesting how the API would fit into existing WPT infra, etc). Feel free to reach out to me (either via GH tagging or by email at smcgruer@chromium.org).

**Comment by jgraham** [comment]
IMO this needs more than an RFC since it's proposing to add a native API in browsers. We can use the RFC as a consensus-building process, but the output of the RFC should be something that looks like a spec rather than a code change in web-platform-tests (although the latter may also be required).

**Comment by stephenmcgruer** [comment]
> IMO this needs more than an RFC since it's proposing to add a native API in browsers. We can use the RFC as a consensus-building process, but the output of the RFC should be something that looks like a spec rather than a code change in web-platform-tests (although the latter may also be required).

That's a fair point. Traditionally our two ways of doing that have been webdriver extensions (established, but can be a difficult process to go through given the many layers involved) or test-only APIs (much newer, no general agreement yet though webxr has cross-browser support I think, but are also much easier to write). Do you think this would follow one of those paths, or are you thinking of a different approach?

**Comment by jgraham** [comment]
I've started https://jgraham.github.io/browser-test/ for APIs like this which don't make sense in WebDriver, can't be exposed to content, but are needed to write tests. Assuming there's interest in progressing that approach I'll figure out how to move the spec somewhere more official so that we can collaborate on it.

In terms of fallback I just don't expect that to be possible for this kind of thing. Generally if a test-required feature's in a spec we don't worry about what happens when you try to run the tests without implementing that spec. I imagine this could be a problem for e.g. running release builds of Safari where it's less likely that there will be a way to turn on test-only APIs. But that may just have to be a problem we live with.

**Comment by domenic** [comment]
That's awesome! /cc @hiroshige-g as it would be a much easier way to unblock our work on import map testing.
